mfs/ops.go

```diff
@@ -46,21 +34,42 @@ func Mv(r *Root, src, dst string) error {
 		return err
 	}
 
-	fsn, err := dstDir.Child(filename)
-	if err == nil {
-		switch n := fsn.(type) {
-		case *File:
-			_ = dstDir.Unlink(filename)
-		case *Directory:
-			dstDir = n
-		default:
-			return fmt.Errorf("unexpected type at path: %s", dst)
+	var dstDirStr string
+	var dstDir *Directory
+	if dst[len(dst)-1] == '/' {
+		dstDirStr = dst
+		dstDir, err = lookupDir(r, dstDirStr)
+		if err != nil {
+			return err
+		}
+		_, err = dstDir.Child(srcFname)
+		if err == nil {
+			return os.ErrExist
+		}
+
+	} else {
+		dstDirStr, dirFileName := gopath.Split(dst)
+		dstDir, err = lookupDir(r, dstDirStr)
+		if err != nil {
+			return err
+		}
+
+		fsn, err := dstDir.Child(dirFileName)
+		if err == nil {
+			switch n := fsn.(type) {
+			case *File:
+				return fmt.Errorf("%s : is not a directory", dst)
+			case *Directory:
+				dstDir = n
```

Comment:
Need to check if there exists a matching file in this directory (replacing it if both the source and dest are files or the source is a directory and the dest is an empty directory).

Reply:
got it, so the `mv` should works as follows ?
```
if src.Type == dir:
     if dest.Type == dir: 
            if desc.Contains(src): case an error
            else move src to dest/
     else if dest.Type == file: case a error
else if src.Type == file:
     if dest.Type == file: replace(src, dest)
     else if desc.Type == directory: 
              if desc.Contains(src): case an error
              else move src to desc/
```

Reply:
Ok, so, ignore what I said about "POSIX". Let's just do what GNU mv does because nobody cares about POSIX (I was about to quote the manual at you and then I actually read the manual; I don't think the GNU people and the POSIX people talk much).

Three modifications:

1. Replacing empty directories is fine.
2. If dest ends in `/`, assert that it is a directory.
3. If src ends in `/`, assert that it is a directory.

```
if src.Type == dir:
     if dest.Type == dir: 
            if dest.Contains(src):
                if dest.Empty(): unlink
                else: return an error
            move src to dest/
     else if dest.Type == file: case a error
else if src.Type == file:
     if dest.Type == file: replace(src, dest)
     else if desc.Type == directory: 
              if desc.Contains(src):
                  if desc.Type == dir and not desc.Empty():
                      return an error
                  unlink src from desc
              move src to desc/
```

(I'm missing the asserts in that pseudo code)

---

Basically, this command should do what the user expects from using linux commands.

Reply:
Eventually, we should probably add the `-T`/`--no-target-directory` flags from Linux for consistency.

The other way to have gone would have been to make everything explicit up-front (mv moves a file to dest exactly, none of this target directory nonsense). However, that'll *really* confuse users and would be a breaking change so... not going to happen.